// @roots/bud-server/lib/cjs/middleware/proxy/url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.URL = void 0;
/**
 * URL helpers for proxy middleware
 *
 * @public
 */
class URL {
    /**
     * Class constructor
     *
     * @public
     */
    constructor(_app) {
        this._app = _app;
    }
    /**
     * Application instance
     *
     * @public
     */
    get app() {
        return this._app();
    }
    /**
     * Server config
     *
     * @public
     */
    get config() {
        return this.app.store.get('server');
    }
    /**
     * Node URL for dev
     *
     * @public
     */
    get dev() {
        return this.config.dev.url;
    }
    /**
     * Node URL for proxy
     *
     * @public
     */
    get proxy() {
        return this.config.proxy.url;
    }
    /**
     * Asset public path
     *
     * @public
     */
    get publicPath() {
        const publicPath = this.app.hooks.filter('build.output.publicPath');
        return publicPath !== 'auto' ? publicPath : '/';
    }
    /**
     * Asset base path for dev
     *
     * @public
     */
    get devAssetUrlBase() {
        return `${this.dev.origin}${this.publicPath}`;
    }
    /**
     * Asset base path for proxy
     *
     * @public
     */
    get proxyAssetUrlBase() {
        return `${this.proxy.origin}${this.publicPath}`;
    }
}
exports.URL = URL;
