// @roots/bud-compiler/lib/cjs/Compiler/compiler.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Compiler = void 0;
const bud_framework_1 = require("@roots/bud-framework");
const bud_support_1 = require("@roots/bud-support");
const webpack_1 = require("webpack");
const budProcess = __importStar(require("./process"));
const { isFunction } = bud_support_1.lodash;
/**
 * Wepback compilation controller class
 *
 * @public
 */
class Compiler extends bud_framework_1.Service {
    constructor() {
        super(...arguments);
        /**
         * Compilation stats
         *
         * @public
         */
        this.stats = {
            assets: [],
            errors: [],
            warnings: [],
        };
        /**
         * Compiler errors
         */
        this.errors = [];
        /**
         * True if compiler is already instantiated
         *
         * @public
         */
        this.isCompiled = false;
        /**
         * @public
         */
        this.config = [];
    }
    /**
     * Initiates compilation
     *
     * @returns the compiler instance
     *
     * @public
     * @decorator `@bind`
     */
    async compile() {
        const config = await this.before();
        const compiler = await this.invoke(config);
        this.app.timeEnd('bud');
        return compiler;
    }
    /**
     * @public
     * @decorator `@bind`
     */
    async invoke(config) {
        config = await this.app.hooks.filterAsync('event.compiler.before', config);
        this.instance = (0, webpack_1.webpack)(this.app.hooks.filter('config.override', config));
        this.instance.hooks.done.tap(config.shift().name, async (stats) => {
            if (this.app.isDevelopment)
                await this.handleStats(stats);
        });
        new webpack_1.ProgressPlugin((...args) => {
            args[1] = args[1].replace('[0] ', '');
            const shouldLog = !this.progress ||
                !this.progress[1] ||
                (args[1] !== this.progress[1] && args[1] !== '');
            this.progress = [args[0] * 100, args[1]];
            this.progress[0] < 100 && shouldLog
                ? budProcess.logger.await(`[${Math.ceil(this.progress[0])}] `, this.progress[1])
                : budProcess.logger.success(`[${Math.ceil(this.progress[0])}] `, `🎉 ${this.progress[1]}`);
        }).apply(this.instance);
        this.app.hooks.filter('event.compiler.after', this.app);
        return this.instance;
    }
    /**
     * Returns final webpack configuration
     *
     * @public
     * @decorator `@bind`
     */
    async before() {
        const config = [];
        this.isCompiled = true;
        await this.app.build.make();
        /**
         * Attempt to use the parent instance in the compilation if there are entries
         * registered to it or if it has no child instances registered.
         */
        if (!this.app.hasChildren) {
            this.app.info(`using config from parent compiler`);
            config.push(this.app.build.config);
            return config;
        }
        this.app.warn(`root compiler will not be tapped (child compilers in use)`);
        /**
         * If there are {@link Framework.children} instances, iterate through
         * them and add to `config`
         */
        await Promise.all(this.app.children?.getValues().map(async (instance) => {
            if (!instance.name)
                return;
            await instance.build.make();
            config.push(instance.build.config);
        }));
        return config;
    }
    async callback(error, stats) {
        await this.handleErrors(error);
        await this.handleStats(stats);
        this.instance.close(err => {
            err && this.app.error(err);
            this.app.close();
        });
    }
    async handleStats(stats) {
        if (!stats?.toJson || !isFunction(stats?.toJson))
            return;
        this.stats = stats.toJson(this.app.store.get('build.stats'));
        budProcess.stats.write(this.stats, this.app.store.get('theme.colors'));
        await this.app.hooks.filter('event.compiler.stats', async () => this.stats);
    }
    async handleErrors(error) {
        if (!error)
            return;
        this.app.isDevelopment &&
            this.app.server.middleware?.hot?.publish({ error });
        this.app.error(error);
    }
}
__decorate([
    bud_support_1.bind,
    bud_support_1.once
], Compiler.prototype, "compile", null);
__decorate([
    bud_support_1.bind
], Compiler.prototype, "invoke", null);
__decorate([
    bud_support_1.bind,
    bud_support_1.once
], Compiler.prototype, "before", null);
__decorate([
    bud_support_1.bind
], Compiler.prototype, "callback", null);
__decorate([
    bud_support_1.bind
], Compiler.prototype, "handleStats", null);
__decorate([
    bud_support_1.bind
], Compiler.prototype, "handleErrors", null);
exports.Compiler = Compiler;
